src/Hedgehog/Tree.fs

```diff
@@ -1,120 +1,105 @@
+        parent |> addChild (singleton value)
 
-    let depth (tree: Tree<'a>) : int =
+    let rec cata f tree =
+        f (root tree) (Seq.map (cata f) (children tree))
+
+    let depth tree =
         tree |> cata (fun _ -> Seq.fold max -1 >> (+) 1)
 
-    let toSeq (tree: Tree<'a>) : 'a seq =
+    let toSeq tree =
         tree |> cata (fun a -> Seq.join >> Seq.cons a)
 
-    /// Map over a tree.
-    let rec map (f : 'a -> 'b) (Node (x, xs) : Tree<'a>) : Tree<'b> =
-        Node (f x, Seq.map (map f) xs)
-
-    let mapWithSubtrees (f: 'a -> seq<Tree<'b>> -> 'b) (tree: Tree<'a>) : Tree<'b> =
-        tree |> cata (fun a subtrees -> Node (f a subtrees, subtrees))
-
-    let rec bind (k : 'a -> Tree<'b>) (Node (x, xs0) : Tree<'a>) : Tree<'b> =
-        match k x with
-        | Node (y, ys) ->
-            let xs = Seq.map (bind k) xs0
-            Node (y, Seq.append xs ys)
-
-    let join (xss : Tree<Tree<'a>>) : Tree<'a> =
-        bind id xss
-
-    /// Turns a tree, in to a tree of trees. Useful for testing Hedgehog itself as
-    /// it allows you to observe the shrinks for a value inside a property,
-    /// while still allowing the property to shrink to a minimal
-    /// counterexample.
-    let rec duplicate (Node (_, ys) as x : Tree<'a>) : Tree<Tree<'a>> =
-        Node (x, Seq.map duplicate ys)
-
-    /// Fold over a tree.
-    let rec fold (f : 'a -> 'x -> 'b) (g : seq<'b> -> 'x) (Node (x, xs) : Tree<'a>) : 'b =
-        f x (foldForest f g xs)
-
-    /// Fold over a list of trees.
-    and foldForest (f : 'a -> 'x -> 'b) (g : seq<'b> -> 'x) (xs : seq<Tree<'a>>) : 'x =
-        Seq.map (fold f g) xs |> g
-
-    /// Build a tree from an unfolding function and a seed value.
-    let rec unfold (f : 'b -> 'a) (g : 'b -> seq<'b>) (x : 'b) : Tree<'a> =
-        Node (f x, unfoldForest f g x)
-
-    /// Build a list of trees from an unfolding function and a seed value.
-    and unfoldForest (f : 'b -> 'a) (g : 'b -> seq<'b>) (x : 'b) : seq<Tree<'a>> =
-        g x |> Seq.map (unfold f g)
-
-    /// Apply an additional unfolding function to an existing tree.
-    ///
-    /// The root outcome remains intact, only the shrinks are affected, this
-    /// applies recursively, so shrinks can only ever be added using this
-    /// function.
-    ///
-    /// If you want to replace the shrinks altogether, try:
-    ///
-    /// Tree.unfold f (outcome oldTree)
-    ///
-    let rec expand (f : 'a -> seq<'a>) (Node (x, xs) : Tree<'a>) : Tree<'a> =
+    let rec map mapping tree =
+        let root = mapping (root tree)
+        let children = Seq.map (map mapping) (children tree)
+        Node (root, children)
+
+    let mapWithSubtrees mapping tree =
+        tree |> cata (fun root subtrees -> Node (mapping root subtrees, subtrees))
+
+    let rec bind mapping tree =
+        let newTree = mapping (root tree)
+        let newChildren = Seq.map (bind mapping) (children tree)
+        Node (root newTree, Seq.append newChildren (children newTree))
```

Comment:
Rename `mapping` to `f` and decompose the tree in the input.

Actually, I think there should be a way to express this as `cata`.  I will look into that.

Reply:
`bind`/`join` is defined in terms of `cata` in [this post](https://blog.ploeh.dk/2019/06/10/tree-catamorphism)...in Haskell, so I need to get some help reading it.  That post also defines `apply` in terms of `cata`, which I will want soon when adding proper applicative support (instead of via `bind`/`join` as is the case now).

Reply:
Wait, it isn't so bad.  It is "just"
```fsharp
let bind f =
  let g oldRoot newOldChildren =
    let Node (newRoot, newNewChildren) = f oldRoot
    Node (newRoot, Seq.concat newNewChildren newOldChildren)
  cata g
```